src/ProducerEventProcessor/UploaderCaptureEventCollectorTest.cpp

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <gmock/gmock.h>
+#include <grpcpp/grpcpp.h>
+#include <grpcpp/support/channel_arguments.h>
+#include <gtest/gtest.h>
+
+#include <thread>
+
+#include "CaptureFile/CaptureFileOutputStream.h"
+#include "OrbitBase/Logging.h"
+#include "ProducerEventProcessor/UploaderCaptureEventCollector.h"
+#include "capture.pb.h"
+#include "services.grpc.pb.h"
+
+using orbit_grpc_protos::ClientCaptureEvent;
+
+namespace orbit_producer_event_processor {
+
+namespace {
+
+static constexpr const char* kAnswerString = "Test Interned String";
+static constexpr uint64_t kAnswerKey = 42;
+
+static ClientCaptureEvent CreateInternedStringCaptureEvent(uint64_t key, const std::string& str) {
+  ClientCaptureEvent event;
+  event.mutable_interned_string()->set_key(key);
+  event.mutable_interned_string()->set_intern(str);
+  return event;
+}
+
+static ClientCaptureEvent CreateCaptureFinishedEvent() {
+  ClientCaptureEvent event;
+  event.mutable_capture_finished()->set_status(orbit_grpc_protos::CaptureFinished::kSuccessful);
+  return event;
+}
+
+class UploaderCaptureEventCollectorTest : public testing::Test {
+ public:
+  void Start(uint64_t event_count, uint64_t produce_duration_ms, uint64_t upload_duration_ms) {
+    event_count_ = event_count;
+    produce_duration_ms_ = produce_duration_ms;
+    upload_duration_ms_ = upload_duration_ms;
+
+    wait_all_data_uploaded_duration_ms_ =
+        upload_duration_ms_ * 2 + produce_duration_ms_ * event_count_;
+
+    data_produce_thread_ =
+        std::thread(&UploaderCaptureEventCollectorTest::ProduceCaptureData, this);
+    data_upload_thread_ = std::thread(&UploaderCaptureEventCollectorTest::UploadCaptureData, this);
```

Comment:
Why is this thread needed? It seems we aren't doing anything with the resulting buffer? Why can't we call `collector_.RefreshUploadDataBuffer` on the main thread (assuming we need to call it at all)?

Reply:
I'm trying to make the test as similar as the real use case of the cloud collector, where we have data producer and data consumer in different threads. Hence, I prefer to keep the upload thread here.